include/evaluators/blas_tree_evaluator_base.hpp

```diff
@@ -0,0 +1,48 @@
+#ifndef BLAS_TREE_EVALUATOR_BASE_HPP_L8QC4HLH
+#define BLAS_TREE_EVALUATOR_BASE_HPP_L8QC4HLH
+
+#include <evaluators/blas_functor_traits.hpp>
+#include <operations/blas_trees.hpp>
+#include <views/operview_base.hpp>
+
+namespace blas {
+namespace internal {
+
+/*! DetectScalar.
+ * @brief Class specialization used to detect scalar values in ScalarOp nodes.
+ * When the value is not an integral basic type,
+ * it is assumed to be a vector and the first value
+ * is used.
+ */
+template <typename T,
+          bool isscal = std::is_same<T, int>::value ||
+                        std::is_same<T, float>::value ||
+                        std::is_same<T, double>::value ||
+                        std::is_same<T, std::complex<float>>::value ||
+                        std::is_same<T, std::complex<double>>::value>
```

Comment:
Because we want to specify explicitly which types are supported by this blas library. I just shortened DetectScalar which was instantiating the same class N times for each primitive type.